container_api/src/container.rs

```diff
@@ -180,12 +182,20 @@ fn create_context(_: &String, path: &String) -> Result<Context, HolochainError>
     let agent_path = format!("{}/state", path);
     create_path_if_not_exists(&cas_path)?;
     create_path_if_not_exists(&eav_path)?;
+    let res = P2pNetwork::new(
```

Comment:
i was going to reject this because of the use of `json!` rather than just `.into()` but then i see what you're working with...

```rust
impl P2pNetwork {
    /// create a new p2p network instance, given message handler and config json
    pub fn new(handler: NetHandler, config: &JsonString) -> NetResult<Self> {
        let config: serde_json::Value = serde_json::from_str(config.into())?;

        // so far, we have only implemented the "ipc" backend type
        match config["backend"].to_string().as_str() {
            "\"ipc\"" => {
                // create a new ipc backend with the passed sub "config" info
                Ok(P2pNetwork {
                    con: NetConnectionThread::new(
                        handler,
                        Box::new(move |h| {
                            let out: Box<NetWorker> = Box::new(IpcNetWorker::new(
                                h,
                                &(config["config"].to_string().into()),
                            )?);
                            Ok(out)
                        }),
                    )?,
                })
            }
            "\"mock\"" => Ok(P2pNetwork {
                con: NetConnectionThread::new(
                    handler,
                    Box::new(move |h| Ok(Box::new(MockWorker::new(h)?) as Box<NetWorker>)),
                )?,
            }),
            _ => bail!("unknown p2p_network backend: {}", config["backend"]),
        }
    }

    /// stop the network module (disconnect any sockets, join any threads, etc)
    pub fn stop(self) -> NetResult<()> {
        self.con.stop()
    }
}
```

looks wrong to me because it is missing a struct/enum for config, so then that missing implementation bleeds into the rest of the codebase :(

i'll approve this work and put a followup

Reply:
@neonphog please review this comment, i assume that you setup this raw serde value handling?

Reply:
@thedavidmeister - my thinking when I set this up was that this config would ultimately come directly from a config file somewhere. If we are going to allow arbitrary plug-able networking modules (processes) we will need to allow arbitrary configuration of them.

I don't feel any particular possessiveness of this code. If you don't like it, feel free to change it.

Reply:
@neonphog i agree. i'm not sure how what you are saying relates to what i said :sweat_smile: 

my thoughts:

if we are manipulating serialized values as though they are a struct, use a real struct
if we want to keep the config structure open to extension then serialized values won't achieve this, so use a trait